SONARJAVA-1972 - S2095 Allow users to specify types to be excluded fr…

fergarrui opened the pull request:
* Added a comma separated list as a parameter to exclude types in the check.

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -72,9 +81,21 @@
   };
+  
+  @Override
+  public void init(MethodTree methodTree, CFG cfg) {
```
Please do not override the `init()` method. Symbolic execution based checks are sensitive, and in this case we don't want to touch anything in the initialization phase.

The best approach in order to retrieve the list of excluded resource types is to use a dedicated (private) getter, which lazily initialize the list (which then should not be static, with all the side effects you noticed).

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -116,34 +137,39 @@ private static boolean needsClosing(Type type) {
         return false;
       }
     }
+    for (String excludedType : excludedTypesList) {
```
If the list is `static`, then you don't have to remove the static modifier everywhere. However, I would prefer you to use a dedicated getter lazily loading the configuration of excluded types.

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -36,27 +44,28 @@
 
+  private static final String DEFAULT_EXCLUDED_TYPES = "";
+  
+  @RuleProperty(
```
I updated the description of the rule in order to include the parameter. Please follow the description from here for keys and description: [RSPEC-3591](https://jira.sonarsource.com/browse/RSPEC-3591)

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -72,9 +81,21 @@
     "org.springframework.context.ConfigurableApplicationContext"
   };
+  
+  private static final List<String> excludedTypesList = new ArrayList<>();
```
Put this list (which should not be static) next to the other rule property-related field. Grouping will help.

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -116,34 +137,39 @@ private static boolean needsClosing(Type type) {
   }
 
-  private static boolean isCloseable(Type type) {
+  private boolean isCloseable(Type type) {
```
I'm not sure you need to remove all of these `static` modifiers. Some of them are probably still valid (this one notably)

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -19,7 +19,15 @@
  */
 package org.sonar.java.se.checks;
 
+import java.util.ArrayList;
```
Please discard these changes, and only limit the modifications to the new imports.

fergarrui commented:
Hi, I did the changes request:
* Discarded the changes in the imports (I'm using Eclipse formatter suggested by Sonar, that's why they were reordered, I'll keep in mind not to reorder them again).
* Rule properties updated to match [RSPEC-3591](https://jira.sonarsource.com/browse/RSPEC-3591)
* Grouped parameter-related properties. Also removed static modifier from the list holding the excluded types.
* Removed init() override and added a lazy load to populate the list.
* Added back some static modifiers, but I can't add most of them.

Thanks for your suggestions and patience! Please let me know if you want more improvements/changes.

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -116,29 +125,43 @@ private static boolean needsClosing(Type type) {
+  }
 
-  private static boolean isCloseable(Type type) {
+  private boolean isCloseable(Type type) {
```
This one can still be static.

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -116,29 +125,43 @@ private static boolean needsClosing(Type type) {
   }
 
-  private static boolean isWithinTryHeader(Tree syntaxNode) {
+  private boolean isWithinTryHeader(Tree syntaxNode) {
```
And when `isTryStatementResource` is static, this one can be static as well.

m-g-sonar commented on java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java:
```diff
@@ -116,29 +125,43 @@ private static boolean needsClosing(Type type) {
   }
 
-  private static boolean isTryStatementResource(VariableTree variable) {
+  private boolean isTryStatementResource(VariableTree variable) {
```
This one can be static.

fergarrui commented:
Changes done :)
